java/algorithm/src/test/java/org/apache/arrow/algorithm/dictionary/TestHashTableBasedDictionaryEncoder.java

```diff
@@ -0,0 +1,166 @@
+  public void testBuildVariableWidthDictionary() {
+    try (VarCharVector vec = new VarCharVector("", allocator);
+         VarCharVector dictionary = new VarCharVector("", allocator)) {
+
+      vec.allocateNew(100, 10);
+      vec.setValueCount(10);
+
+      dictionary.allocateNew();
+
+      // fill data
+      vec.set(0, "hello".getBytes());
+      vec.set(1, "abc".getBytes());
+      vec.setNull(2);
+      vec.set(3, "world".getBytes());
+      vec.set(4, "12".getBytes());
+      vec.set(5, "dictionary".getBytes());
+      vec.setNull(6);
+      vec.set(7, "hello".getBytes());
+      vec.set(8, "good".getBytes());
+      vec.set(9, "abc".getBytes());
+
+      HashTableBasedDictionaryBuilder<VarCharVector> dictionaryBuilder =
+              new HashTableBasedDictionaryBuilder<>(dictionary, encodeNull);
+
+      int result = dictionaryBuilder.addValues(vec);
+
+      if (encodeNull) {
+        assertEquals(7, result);
+        assertEquals(7, dictionary.getValueCount());
+
+        assertEquals("hello", new String(dictionary.get(0)));
+        assertEquals("abc", new String(dictionary.get(1)));
+        assertNull(dictionary.get(2));
+        assertEquals("world", new String(dictionary.get(3)));
+        assertEquals("12", new String(dictionary.get(4)));
+        assertEquals("dictionary", new String(dictionary.get(5)));
+        assertEquals("good", new String(dictionary.get(6)));
+      } else {
+        assertEquals(6, result);
+        assertEquals(6, dictionary.getValueCount());
+
+        assertEquals("hello", new String(dictionary.get(0)));
+        assertEquals("abc", new String(dictionary.get(1)));
+        assertEquals("world", new String(dictionary.get(2)));
+        assertEquals("12", new String(dictionary.get(3)));
+        assertEquals("dictionary", new String(dictionary.get(4)));
+        assertEquals("good", new String(dictionary.get(5)));
+      }
+    }
+  }
+
+  @Test
+  public void testBuildFixedWidthDictionary() {
+    try (IntVector vec = new IntVector("", allocator);
+         IntVector dictionary = new IntVector("", allocator)) {
+      vec.allocateNew(10);
+      vec.setValueCount(10);
+
+      dictionary.allocateNew();
+
+      // fill data
+      vec.set(0, 4);
+      vec.set(1, 8);
+      vec.set(2, 32);
+      vec.set(3, 8);
+      vec.set(4, 16);
+      vec.set(5, 32);
+      vec.setNull(6);
+      vec.set(7, 4);
+      vec.set(8, 4);
+      vec.setNull(9);
+
+      HashTableBasedDictionaryBuilder<IntVector> dictionaryBuilder =
+              new HashTableBasedDictionaryBuilder<>(dictionary, encodeNull);
+
+      int result = dictionaryBuilder.addValues(vec);
+
+      if (encodeNull) {
+        assertEquals(5, result);
+        assertEquals(5, dictionary.getValueCount());
+
+
+        assertEquals(4, dictionary.get(0));
+        assertEquals(8, dictionary.get(1));
+        assertEquals(32, dictionary.get(2));
+        assertEquals(16, dictionary.get(3));
+        assertTrue(dictionary.isNull(4));
+      } else {
```

Comment:
see comment on other CL, we should either split up the test or parameterize expected values as well.

Reply:
sure. tests split and javadoc updated.